File: iosApp/iosApp/Modules/Account/AccountViewModel.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

struct AccountViewModel {
    let currencyName: String
    let header: AccountViewModel.Header
    let address: AddressViewModel
    let candles: CandlesViewModel
    let marketInfo: AccountViewModel.MarketInfo
    let bonusAction: AccountViewModel.BonusAction?
    let transactions: [AccountViewModel.Transaction]
}

extension AccountViewModel {

    struct Header {
        let balance: [Formatters.Output]
        let fiatBalance: String
        let pct: String
        let pctUp: Bool
        let buttons: [Button]

        struct Button {
            let title: String
            let imageName: String
        }
    }
}

extension AccountViewModel {

    struct AddressViewModel {
        let address: String
        let copyIcon: String
    }
}

extension AccountViewModel {

    struct MarketInfo {
        let marketCap: String
        let price: String
        let volume: String
    }
}

extension AccountViewModel {

    struct BonusAction {
        let title: String
    }
}

extension AccountViewModel {

    enum Transaction {
        case empty(text: String)
        case loading(text: String)
        case data(Data)
        
        struct Data {
            let date: String
            let address: String
            let amount: String
            let isReceive: Bool
            let txHash: String
        }
        
        var empty: String? {
            switch self {
            case let .empty(text): return text
            default: return nil
            }
        }
        
        var loading: String? {
            switch self {
            case let .loading(text): return text
            default: return nil
            }
        }
        
        var data: Data? {
            switch self {
            case let .data(data): return data
            default: return nil
            }
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/account/AccountViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.account

import com.sonsofcrypto.web3lib.formatters.Formater
import com.sonsofcrypto.web3walletcore.common.viewModels.CandlesViewModel

data class AccountViewModel(
    val currencyName: String,
    val header: Header,
    val address: Address,
    val candles: CandlesViewModel,
    val marketInfo: MarketInfo,
    val bonusAction: BonusAction?,
    val transactions: List<Transaction>,
) {
    data class Header(
        val balance: List<Formater.Output>,
        val fiatBalance: List<Formater.Output>,
        val pct: String,
        val pctUp: Boolean,
        val buttons: List<Button>,
    ) {
        data class Button(
            val title: String,
            val imageName: String,
        )
    }

    data class Address(
        val addressFormatted: String,
        val address: String,
    )

    data class MarketInfo(
        val marketCap: List<Formater.Output>,
        val price: List<Formater.Output>,
        val volume: List<Formater.Output>,
    )

    data class BonusAction(
        val title: String,
    )

    sealed class Transaction {
        data class Empty(val text: String): Transaction()
        data class Loading(val text: String): Transaction()
        data class Loaded(val data: Data): Transaction()

        data class Data(
            val date: String,
            val address: String,
            val amount: List<Formater.Output>,
            val fiatPrice: List<Formater.Output>,
            val isReceive: Boolean,
            val txHash: String,
        )
    }
}
